### PR TITLE
[FLINK-15355][plugins]Classloader restricts access to parent to whitelist.

### PR DESCRIPTION
**Pull request by AHeise**
## What is the purpose of the change

Classloader restricts access to parent to whitelist.

No class/resource in the system class loader (everything in lib/) can be seen in the plugin except those starting with a whitelist prefix.

## Brief change log

- Classloader restricts access to parent to whitelist.

## Verifying this change

- `Shaded Hadoop S3A with credentials provider end-to-end test`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (**yes** / no / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by AHeise** [comment]
@flinkbot run travis

**Comment by AHeise** [comment]
@flinkbot run azure

**Comment by AHeise** [comment]
Two things, a) the first commit could be extended all the way to the configuration. b) it’s a bit cumbersome to extend the `ChildFirstClassLoader`, it could be a separate classloader, but we would replicate the parent lookup.

File: flink-core/src/main/java/org/apache/flink/configuration/CoreOptions.java [hunk 1]
```diff
@@ -121,7 +121,7 @@
 	public static final ConfigOption<String> PLUGIN_ALWAYS_PARENT_FIRST_LOADER_PATTERNS = ConfigOptions
 		.key("plugin.classloader.parent-first-patterns.default")
 		.stringType()
-		.defaultValue("java.;scala.;org.apache.flink.;javax.annotation.;org.slf4j;org.apache.log4j;org.apache" +
+		.defaultValue("java.;javax.;scala.;org.apache.flink.;org.slf4j;org.apache.log4j;org.apache" +
```
Comment (pnowojski) [inline flink-core/src/main/java/org/apache/flink/configuration/CoreOptions.java]: revert?

File: flink-core/src/main/java/org/apache/flink/core/plugin/PluginLoader.java [hunk 2]
```diff
@@ -100,4 +105,97 @@ public P next() {
...
+				return flinkClassLoader.getResource(name);
+			}
+
+			URL urlClassLoaderResource = findResource(name);
+
+			if (urlClassLoaderResource != null) {
+				return urlClassLoaderResource;
+			}
+
+			return null;
```
Comment (pnowojski) [inline flink-core/src/main/java/org/apache/flink/core/plugin/PluginLoader.java]:
```
return super.getResource(name);
```
?

File: flink-tests/src/test/java/org/apache/flink/test/plugin/PluginLoaderTest.java [hunk 3]
```diff
@@ -21,6 +21,7 @@
 import org.apache.flink.core.plugin.PluginDescriptor;
 import org.apache.flink.core.plugin.PluginLoader;
 import org.apache.flink.test.plugin.jar.plugina.TestServiceA;
+import org.apache.flink.test.plugin.spi.TestSpi;
```
Comment (pnowojski) [inline flink-tests/src/test/java/org/apache/flink/test/plugin/PluginLoaderTest.java]: Why have all the tests changed in this commit?

**Pull request merged**